lib/plugins/aws/package/compile/functions.js

```diff
@@ -219,9 +218,16 @@ class AwsCompileFunctions {
         ? this.serverless.service.package.deploymentBucket
         : { Ref: 'ServerlessDeploymentBucket' };
 
-      functionResource.Properties.Code.S3Key = `${
-        this.serverless.service.package.artifactDirectoryName
-      }/${artifactFilePath.split(path.sep).pop()}`;
+      if (artifactFilePath.includes(`.serverless${path.sep}`)) {
```

Comment:
I think we could avoid that if statement here. What if we would do something like this (variable names are probably not the best below, I just want to illustrate the approach):

```
const pathWithoutServerless = artifactFilePath.split('.serverless').pop()
```

The above will ensure that for provided paths, it will be assigned to `pathWithoutServerless` and for situations where the artifact from `.serverless` is used, we'll strip it and only use the part of path inside `.serverless` dir. I think we don't need to include `path.sep` in the split because we can uniformly handle that later.

```
const pathWithUnixSlashes = pathWithoutServerless.replace(/\\/g, '/');
const pathWithoutLeadingSlash = pathWithUnixSlashes.startswith('/') ? pathWithUnixSlashes.substring(1) : pathWithUnixSlashes;
```

Here we will have the path with replaced slashes + removed leading slash if it exists. However after second though I guess it might break for paths on Windows like `C:\\test\\test` - does the approach with replacing `path.parse(artifactFilePath`.root)` resolve replacing the `C:\\` ?

What do you think about the above approach? I think it's a bit simplified vs the one that uses `if` branching.

Reply:
It'd break S3Keys on Windows for sure. That's why I started tinkering with `path.parse` to have a unified method for removing a path root and only this, since I don't see any other way to meet our goal and ensure unique S3Key for custom artifacts, especially with absolute path provided.

I was trying to save a path in the Unix standard to prevent playing with regular expressions, without luck so far.
I guess we might try the suggested solution and ensure a path root is removed (via `path.parse`).
I'll mess around and we'll see how it went.

Reply:
Thank you @scadu :bow:

Reply:
I just pushed the latest changes and it seems to be working as expected (tested on Linux and Windows). 
Now I'm just trying to get tests working again, without creating much boilerplate. 
I'll keep the PR as a draft until I figure it out if that's okay with you.

Reply:
Sure thing - thank you for all your work on this one  :bow:

Reply:
Small update: I still have a few tests failing (three - to be precise). I'll keep working on fixing those.

Reply:
Thanks for an update @scadu :bow:

Reply:
It seems that I've hit a dead-end again. I've managed to get most of the tests working without creating much boilerplate, however, I'm still struggling with three of them:

```
  1) AwsCompileFunctions
       #downloadPackageArtifacts()
         should download the file and replace the artifact path for function packages:
      AssertionError: expected 'Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/9350ee01b157468c/new-service.zip' to equal 's3-hosted-artifact.zip'
      + expected - actual

      -Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/9350ee01b157468c/new-service.zip
      +s3-hosted-artifact.zip

      at C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js:123:37
 ```
      
 ```
   2) AwsCompileFunctions
       #compileFunctions()
         should use function artifact if individually:

      AssertionError: expected 'somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/test.zip' to deeply equal 'somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/new-service.zip'
      + expected - actual

      -somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/test.zip
      +somedir/Users/ljend/AppData/Local/Temp/tmpdirs-serverless/0ce6/141a2ea711785880/new-service.zip

      at C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js:170:64
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

```
  3) AwsCompileFunctions
       #compileFunctions()
         should honour provider.iam.role option when set:
     ServerlessError: Serverless command "C:\Users\ljend\Code\serverless\test\unit\lib\plugins\aws\package\compile\functions.test.js" not found. Did you mean "credentials"? Run "serverless help" for a list of all available commands.
      at commandOrAlias.reduce.commands (lib\classes\PluginManager.js:459:15)
      at Array.reduce (<anonymous>)
      at PluginManager.getCommand (lib\classes\PluginManager.js:422:27)
      at PluginManager.validateCommand (lib\classes\PluginManager.js:549:10)
      at Serverless.run (lib\Serverless.js:234:24)
      at C:\Users\ljend\Code\serverless\node_modules\@serverless\test\run-serverless.js:261:36
      at C:\Users\ljend\Code\serverless\node_modules\@serverless\test\setup-run-serverless-fixtures-engine.js:33:22
      at Context.<anonymous> (test\unit\lib\plugins\aws\package\compile\functions.test.js:245:30)
```

For the rest of the tests, [this conditional works perfectly](https://github.com/serverless/serverless/pull/8891/files#diff-b0a2f2c747c874b3ac996b289739e323313d074cf998a8cf83910d2d9380b181R76-R92).
The first test failing has specified `awsCompileFunctions.serverless.service.functions[functionName].package.artifact` which is some hint (though I'm not sure how to pass it properly), but I have no clue about the second one (`should use function artifact if individually`), not to mention the last one.
It looks like it's a small thing that I can't spot.

Reply:
Sorry for delayed response - where do you see these failing tests? I've just checked the last pipeline and it seems like only one of them is failing?

As for the adjustments - if we're going this far with changing the tests, I'm afraid that we should consider replacing them with `runServerless`-based ones which will heavily impact the scope of this PR - from what I see the current ones are quite fragile and I'm not sure about their reliability to be honest. What do you think?

Reply:
One of them is failing [for unit test ran on Node 14](https://github.com/serverless/serverless/pull/8891/checks?check_run_id=2024336592), another two can be seen in the step of [isolated unit test ran on Node 12](https://github.com/serverless/serverless/pull/8891/checks?check_run_id=2024336388).

Regarding reliability, do you mean the conditionals introduced in this PR or the fact that `runServerless` is not used everywhere in `functions.test.js`? Nevertheless, I'll try to include #8523 into this PR.

Reply:
Thanks, I've missed the fact that they're inconsistently failing across different Node versions. 

As for reliability - I'm talking more about the genenral approach used in existing tests - they heavily depend on manually constructing fragile examples for tests which makes it then hard to refactor. Ideally would be to first refactor the package-related tests in `functions.test.js` (not necesarilly all of them because it's a lot of work) to use `runServerless` and then introduce that change - we made a decision to not introduce extra tests with old approach and I feel like with the conditionals and changes here we're kinda adding to "old" tech debt of our testing suite. Please let me know if you're up for that work, as I know it grows the initial scope of the PR and I totally understand if you're not interested.

Reply:
Yeah, that totally makes sense. I just thought that I'd finish this one and pick #8523 but now it seems that the initial @medikoo's suggestion was more reasonable (back then we also had some edge cases to cover).

If it's not urgent (and assuming that the issue is opened for over three years now it isn't but I don't have the same context), I'd take a closer look into rewriting related tests in `functions.test.js` to use `runServerless`, especially when we got this far already.
@pgrzesik let me know what's your position on this issue.

Reply:
That would be perfect @scadu as I think to wrap this one up, we will need updated test suite and I believe it makes a lot of sense to refactor it as a part of separate PR. This one definitely can wait :+1: Thank you for your efforts and help @scadu :bow:

Reply:
Sorry for not being clear - My thinking was that we could put this PR on a pause for a moment, cover necessary parts from #8523 as a part of a separate PR, merge that PR and then come back to this one, relying on new, `runServerless`-based tests only. I know it prolongs this PR, but if you're okay with it, I believe that would be 'cleanest' approach. Let me know that do you think :raised_hands: Thanks for the kind words as well :rocket:

Reply:
@pgrzesik oh, right. I'll update #8523 with this information to make prevent duplicated work.

Reply:
Perfect, thank you :clap: